src/Tools/Common/ReversedServerHelpers/ReversedServerHelpers.cs

```diff
@@ -97,11 +98,12 @@ public bool Start( string diagnosticTransportName, CancellationToken ct, bool sh
                     Console.WriteLine($"Launching: {_childProc.StartInfo.FileName} {_childProc.StartInfo.Arguments}");
                 }
                 _childProc.Start();
+                _processStarted = true;
             }
             catch (Exception e)
             {
                 Console.WriteLine($"Cannot start target process: {_childProc.StartInfo.FileName} {_childProc.StartInfo.Arguments}");
-                Console.WriteLine(e.ToString());
+                Console.WriteLine(e.Message);
```

Comment:
Rather than letting the code handle the error here we can refactor this slightly to get more useful error handling behavior.

1. Move the CommandLineErrorException to the Tools/Common code and remove the "Counters" from
the namespace. This type can be generally useful and has nothing specific to counters.
https://github.dev/dotnet/diagnostics/blob/main/src/Tools/dotnet-counters/CommandLineErrorException.cs

2. Instead of having ProcessLauncher print to the console, let it throw CommandLineErrorException instead and put the text it would have written to the console in the Exception message.

3. Then add a catch clause for the CommandLineErrorException that will print out that message in DiagnosticsServerRouterRunner.runRouter(...) and CollectCommandHandler.Collect(...). For example in Collect(...)

```C#
// these are errors we anticipated and we know the error message will make sense to
// print to the console
catch(CommandLineErrorException e)
{
    Console.Error.WriteLine("[ERROR] " + e.Message);
    collectionStopped = true;
    ret = ReturnCode.TracingError;
}
// these errors could be anything and the messages may not make sense without context
// add the stack trace to help diagnose even though its ugly. Ideally no errors would
// get here.
catch(Exception e)
{
    Console.Error.WriteLine(e.ToString());
    collectionStopped = true;
    ret = ReturnCode.TracingError;
}
```

Give that a shot and see if it produces a better looking result?

Reply:
I did this for CollectCommandHandler.Collect() and the output would look like this:
```
No profile or providers specified, defaulting to trace profile 'cpu-sampling'

Provider Name                           Keywords            Level               Enabled By
Microsoft-DotNETCore-SampleProfiler     0x0000F00000000000  Informational(4)    --profile
Microsoft-Windows-DotNETRuntime         0x00000014C14FCCBD  Informational(4)    --profile

Launching: garbage
[ERROR] Cannot start target process: garbage
```

I'm still not sure though why would add a catch clause for a CommandLineErrorException in DiagnosticsServerRouterRunner.runRouter(). I don't think Process.Start() ever gets called here. DiagnosticsServerRouterRunner.Start() and Router.Start() get called, both of which don't throw a CommandLineException.

Reply:
I was under the impression that runRouter() calls OnRouterStarted() ([here](https://github.com/dotnet/diagnostics/blob/3128b544e9d7241146b7398f93a4f5643865e0fd/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouterRunner.cs#L67))  and OnRouterStarted calls ProcessLauncher.Start() ([here](https://github.com/dotnet/diagnostics/blob/21ea2d84f94a18cc95f0f461a5e41623aacc83d5/src/Tools/dotnet-dsrouter/DiagnosticsServerRouterCommands.cs#L36))

If I got that wrong my underlying intent is if ds-router calls ProcessLauncher.Start() anywhere then we should figure out what is the callstack where that call occurs and make sure the CommandLineErrorException will be caught.

Reply:
Can we also add this to the CommandLineErrorException message? "An error occurred trying to start process 'garbage' with working directory 'C:\code\diagnostics\artifacts\bin\dotnet-trace\Debug\netcoreapp3.1'. The system cannot find the file specified. I assume this part of the text is coming from the Exception.Message property on the exception that is thrown?

Reply:
I see, I hadn't checked OnRouterStarted() and I see the Process.Start() call now. I'll add the catch clause to runRouter().

Also, the message property from the exception thrown by Process.Start() is actually only:
```
The system cannot find the file specified.
```
Don't know if we still want to add it to the CommandLineErrorException message.

Reply:
Also, do you think we really need a separate catch statement for CommandLineErrorExceptions for runRouter()? The catch statement that is currently in place only sends the exception message for all exceptions. Unless we want to change it to log only the message for CommandLineErrorExceptions and the entire ToString() for all other exceptions.

Reply:
@davmason - I've gotten tied up with some 1st party conf stuff and feel bad having the discussion delayed on me. I'm happy for you and Daniel to make the call together. Worst case we can always can always tweak the behavior later.

Reply:
@dramos020 I think we should do what Noah suggested a few comments back and take the exception message from Process.Start and wrap it in a little extra text:

```
throw new CommandLineErrorException($"An error occurred trying to start process '{processName}' with working directory '{workingDir}'. {orginalException.Message}.");
```

Then we don't have to have to add a new handler for runRouter

Reply:
The error message now looks like this:
```
PS C:\code\diagnostics\artifacts\bin\dotnet-trace\Debug\netcoreapp3.1> ./dotnet-trace collect -- garbage
No profile or providers specified, defaulting to trace profile 'cpu-sampling'

Provider Name                           Keywords            Level               Enabled By
Microsoft-DotNETCore-SampleProfiler     0x0000F00000000000  Informational(4)    --profile
Microsoft-Windows-DotNETRuntime         0x00000014C14FCCBD  Informational(4)    --profile

Launching: garbage
[ERROR] An error occured trying to start process garbage with working directory C:\code\diagnostics\artifacts\bin\dotnet-trace\Debug\netcoreapp3.1. The system cannot find the file specified.
```

Reply:
Looks great! 👍